design/026-requirements_traits.md

```diff
@@ -0,0 +1,125 @@
+- ``transitive_libs`` (default None): The libraries to link with of the dependency will be visible downstream or not. The default None allows Conan to auto detect this trait, for example, if the current package is a ``header-only`` one, and it depends on another library (header only, or static/shared), the libraries of the transitive dependency must be available and used in the ``-I<libs>`` and ``-L<libdirs>`` compilation downstream.
+- ``test`` (default False): this requirement is a test library or framework, like Catch2 or gtest. It is mostly a library that needs to be included and linked, but that will not be propagated downstream.
+- ``package_id`` (default None): if the recipe wants to specify how the dependency version affects the current package ``package_id``, can be directly specified here. While it could be also done in the ``package_id()`` method, it seems simpler to be able to specify it in the ``requires`` while avoiding some ambiguities.
+- ``force`` (default False): This ``requires`` will force its version in the dependency graph upstream, overriding other existing versions even of transitive dependencies, and also solving potential existing conflicts.
+- ``override`` (default False): The same as the ``force`` trait, but not adding a ``direct`` dependency. If there is no transitive dependency to override, this ``require`` will be discarded. This trait only exists at the time of defining a ``requires``, but it will not exist as an actual ``requires`` once the graph is fully evaluated
+- ``direct`` (default True): If the dependency is a direct one, that is, it has explicitly been declared by the current recipe, or if it is a transitive one.
+
+Note: In most cases, for normal library requirements, it won’t be necessary to specify any special trait, just using the default ``self.requires(ref)`` or ``requires = "ref"`` will be enough, as the default traits, working together with the package type (shared/static/header-only) can automatically deduce and propagate correctly the information.
+
+The requires and actual dependencies those requires are pointing to are accessible from recipes and tools like:
+
+```python
+def generate(self):
+      for require, dependency in self.dependencies.items():
+           require.direct # boolean
+           require.headers # boolean
+           require.libs # boolean
+           require.build # boolean
+           require.run # boolean
+```
+
+``build_requires`` will remain to be a high-level definition of requirements, but internally it is implemented as:
+
+```python
+req = Requirement(ref, headers=False, libs=False, build=True, run=True, visible=False,
```

Comment:
Reminded me flatbuffers (aka tool + lib):

In this case either duplicate requirements should be allowed or there should be a way to specify that a requirement is for both host and build contexts and will be duplicated internally.

If same requirement can be added multiple times it will be a tough job (IMHO) to specify all possible cases or combinations. Then why not to add with more than one combination of traits?

The second case is better, but then it limits possibility to use different versions in different contexts (maybe it's not bad). In this case a parameter `context: enum or string` fits better than `build: bool`.